benchmarks/nemo/nemolite2d/manual_versions/psykal_c_opencl/time_step.c

```diff
@@ -254,138 +225,221 @@ void c_invoke_time_step(
             check_status("clCreateKernel", ret);
         } 
   
-        /* Create Device Memory Buffers*/
+        /* Create Device Memory Buffers, to make them permanent after the function
```

Comment:
Annoyingly github is not letting me continue our previous conversation. When I said "Could we now do away with all of the xxx_device variables.." I was simply talking about tidying up a bit. If this new implementation is significantly better than the old one then I'm happy. If it has just ended up adding complexity in order to make things look nicer then that's not so good. I have to admit I don't really understand what's going on now! It looks like you have to create a buffer to store each pointer location? Is that so their locations remain resident on the device(between invokes) rather than storing them on the host? Please could you extend the comment a bit more (apologies if I'm being dense).

Reply:
I agree is a bit more complex than before (this is due to C not having references like in C++) but I already wanted to do this before to avoid using globals to store the data. Note that globals would be problematic for multiple invokes as there would be variable duplication to store the fields in each invoke, so PSyclone needs to generate something similar to this current implementation.

The clCreateBuffer creates the buffer in the device and returns an object of metadata in the host that tells the buffer location+some properties. What I do is a copy of this metadata object to the heap (malloc and memcpy) so that is not deleted at the end of the scope. Finally the parameter pointer points to this new location so that it can be reused next time.

Added a comment about this.

Reply:
Great. I think I get it now.